sdk/go/pulumi/resource.go

```diff
@@ -314,7 +316,69 @@ func DeleteBeforeReplace(o bool) ResourceOption {
 // DependsOn is an optional array of explicit dependencies on other resources.
 func DependsOn(o []Resource) ResourceOption {
 	return resourceOption(func(ro *resourceOptions) {
-		ro.DependsOn = append(ro.DependsOn, o...)
+		ro.DependsOn = append(ro.DependsOn, func(ctx context.Context) ([]URN, error) {
+			var urns []URN
+			for _, res := range o {
+				directUrn, isKnown, _, err := res.URN().awaitURN(ctx)
+				if err != nil {
+					return nil, err
+				}
+				if isKnown {
+					urns = append(urns, directUrn)
+				}
+			}
+			return distinctURNs(urns), nil
+		})
+	})
+}
+
+// This function is similar to `DependsOn` but accepts a
+// `ResourceInput` slice instead, allowing to pass `ResourceOutput`
+// values.
+func DependsOnInputs(o []ResourceInput) ResourceOption {
+	return resourceOption(func(ro *resourceOptions) {
+		ro.DependsOn = append(ro.DependsOn, func(ctx context.Context) ([]URN, error) {
+			return resolveDependsOnURNs(ctx, o)
+		})
+	})
+}
+
+// A fully dynamic version of `DependsOn` that accepts `AnyOutput`
+// carrying `[]Resource`, `[]ResourceOutput`, or `[]ResourceInput`
+// values, which are then considered as explicit dependencies.
+func DependsOnOutput(any AnyOutput) ResourceOption {
```

Comment:
Including to be compatible in power with Node/Py SDK where a list of prompty-indeterminate length can be passed in. If we don't care about that, can remove. It's a bit of an unfortunate interface.